### PR TITLE
[SPARK-28058][DOC] Add a note to doc of mode of CSV for column pruning

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

When using `DROPMALFORMED` mode, corrupted records aren't dropped if malformed columns aren't read. This behavior is due to CSV parser column pruning. Current doc of `DROPMALFORMED` doesn't mention the effect of column pruning. Users will be confused by the fact that `DROPMALFORMED` mode doesn't work as expected.

Column pruning also affects other modes. This is a doc improvement to add a note to doc of `mode` to explain it.

## How was this patch tested?

N/A. This is just doc change.

### CODE REVIEW

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -441,7 +441,12 @@ def csv(self, path, schema=None, sep=None, encoding=None, quote=None, escape=Non
                   When it meets a record having fewer tokens than the length of the schema, \
                   sets ``null`` to extra fields. When the record has more tokens than the \
                   length of the schema, it drops extra tokens.
-                * ``DROPMALFORMED`` : ignores the whole corrupted records.
+                * ``DROPMALFORMED`` : ignores the whole corrupted records. Note that when CSV \
+                  parser column pruning (``spark.sql.csv.parser.columnPruning.enabled``) is enabled \
```
Comment (viirya) [inline python/pyspark/sql/readwriter.py]: One concern is that `spark.sql.csv.parser.columnPruning.enabled` is internal config. I'm not sure if there is policy that we shouldn't mention internal config in doc.

cc @HyukjinKwon @dongjoon-hyun

File: python/pyspark/sql/readwriter.py [hunk 2]
```diff
@@ -441,7 +441,12 @@ def csv(self, path, schema=None, sep=None, encoding=None, quote=None, escape=Non
                   When it meets a record having fewer tokens than the length of the schema, \
                   sets ``null`` to extra fields. When the record has more tokens than the \
                   length of the schema, it drops extra tokens.
-                * ``DROPMALFORMED`` : ignores the whole corrupted records.
+                * ``DROPMALFORMED`` : ignores the whole corrupted records. Note that when CSV \
+                  parser column pruning (``spark.sql.csv.parser.columnPruning.enabled``) is \
+                  enabled (it is enabled by default), the malformed columns can be ignored during \
+                  parsing if they are pruned, resulting the corrupted records are not dropped. \
+                  Disabling the column pruning feature can drop corrupted records even malformed \
+                  columns are not read.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: I think it's good to note. But for JIRA itself, do you know the cause? I haven't taken a close look but it looked to me like Univocity parses the tokens weirdly in a certain condition. It is possible that it's a bug.

Comment (HyukjinKwon) [reply python/pyspark/sql/readwriter.py]: If we're going to investigate it separately, let's just file a separate JIRA for this note. Then it should be good to go orthogonally.

Comment (viirya) [reply python/pyspark/sql/readwriter.py]: Not sure which one you think is possibly a bug. Let's discuss it on the JIRA?

Comment (viirya) [reply python/pyspark/sql/readwriter.py]: > If we're going to investigate it separately, let's just file a separate JIRA for this note. Then it should be good to go orthogonally.

Ok. Sounds good.

Comment (HyukjinKwon) [reply python/pyspark/sql/readwriter.py]: Yea, let's discuss there. For this note, looks good and useful to have. I will take another look tomorrow and get this in

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Okie, looks your first judgement was correct. Let me switch the JIRA back.

**Comment by viirya** [comment]
@HyukjinKwon Thanks for spending time on investigating and confirming.

File: python/pyspark/sql/readwriter.py [hunk 3]
```diff
@@ -441,7 +441,12 @@ def csv(self, path, schema=None, sep=None, encoding=None, quote=None, escape=Non
                   When it meets a record having fewer tokens than the length of the schema, \
                   sets ``null`` to extra fields. When the record has more tokens than the \
                   length of the schema, it drops extra tokens.
-                * ``DROPMALFORMED`` : ignores the whole corrupted records.
+                * ``DROPMALFORMED`` : ignores the whole corrupted records. Note that when CSV \
+                  parser column pruning (``spark.sql.csv.parser.columnPruning.enabled``) is \
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: Actually, @viirya, I think this note applies to all other modes. Would it be better to leave a short node under `mode` parameter? I think we can describe, for instance ..

Note that Spark tries to parse only required columns in CSV. Therefore, malformed records can be different based on required set of fields. This behavior can be controlled by ``spark.sql.csv.parser.columnPruning.enabled`` (enabled by default)

Feel free to reword this ..

Comment (HyukjinKwon) [reply python/pyspark/sql/readwriter.py]: Here is what I tested to show the diff by that configuration and parse mode:

PERMISSIVE:

```scala
scala> spark.conf.set("spark.sql.csv.parser.columnPruning.enabled", "true")

scala> spark.read.schema("fruit string, color string, price string, quantity string, _corrupt_record string").option("header", "true").csv("fruit.csv").select('fruit, 'color, '_corrupt_record).show()
+------+------+---------------+
| fruit| color|_corrupt_record|
+------+------+---------------+
| apple|   red|           null|
|banana|yellow|           null|
|orange|orange|           null|
|   xxx|  null|           null|
+------+------+---------------+


scala> spark.conf.set("spark.sql.csv.parser.columnPruning.enabled", "false")

scala> spark.read.schema("fruit string, color string, price string, quantity string, _corrupt_record string").option("header", "true").csv("fruit.csv").select('fruit, 'color, '_corrupt_record).show()
+------+------+---------------+
| fruit| color|_corrupt_record|
+------+------+---------------+
| apple|   red|           null|
|banana|yellow|           null|
|orange|orange|           null|
|   xxx|  null|            xxx|
+------+------+---------------+
```

DROPMALFORMED:

```scala
scala> spark.conf.set("spark.sql.csv.parser.columnPruning.enabled", "false")

scala> spark.read.option("header", "true").option("mode", "DROPMALFORMED").csv("fruit.csv").select('fruit, 'color).show()
+------+------+
| fruit| color|
+------+------+
| apple|   red|
|banana|yellow|
|orange|orange|
+------+------+


scala> spark.conf.set("spark.sql.csv.parser.columnPruning.enabled", "true")

scala> spark.read.option("header", "true").option("mode", "DROPMALFORMED").csv("fruit.csv").select('fruit, 'color).show()
+------+------+
| fruit| color|
+------+------+
| apple|   red|
|banana|yellow|
|orange|orange|
|   xxx|  null|
+------+------+
```

FAILFAST:

```scala
scala> spark.conf.set("spark.sql.csv.parser.columnPruning.enabled", "false")

scala> spark.read.option("header", "true").option("mode", "FAILFAST").csv("fruit.csv").select('fruit, 'color).show()
org.apache.spark.SparkException: Malformed records are detected in record parsing. Parse Mode: FAILFAST. To process malformed records as null result, try setting the option 'mode' as 'PERMISSIVE'.
	at org.apache.spark.sql.catalyst.util.FailureSafeParser.parse(FailureSafeParser.scala:70)
	at org.apache.spark.sql.catalyst.csv.UnivocityParser$.$anonfun$parseIterator$2(UnivocityParser.scala:347)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:484)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:490)
	at org.apache.spark.sql.execution.datasources.v2.PartitionReaderFromIterator.next(PartitionReaderFromIterator.scala:26)
	at org.apache.spark.sql.execution.datasources.v2.PartitionReaderWithPartitionValues.next(PartitionReaderWithPartitionValues.scala:48)
	at org.apache.spark.sql.execution.datasources.v2.PartitionedFileReader.next(FilePartitionReaderFactory.scala:55)
	at org.apache.spark.sql.execution.datasources.v2.FilePartitionReader.next(FilePartitionReader.scala:70)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.hasNext(DataSourceRDD.scala:62)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:701)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:292)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:852)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:852)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:327)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:291)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:125)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:426)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1350)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:429)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.spark.sql.catalyst.util.BadRecordException: java.lang.RuntimeException: Malformed CSV record
	at org.apache.spark.sql.catalyst.csv.UnivocityParser.org$apache$spark$sql$catalyst$csv$UnivocityParser$$convert(UnivocityParser.scala:238)
	at org.apache.spark.sql.catalyst.csv.UnivocityParser.$anonfun$doParse$1(UnivocityParser.scala:192)
	at org.apache.spark.sql.catalyst.csv.UnivocityParser.parse(UnivocityParser.scala:203)
	at org.apache.spark.sql.catalyst.csv.UnivocityParser$.$anonfun$parseIterator$1(UnivocityParser.scala:343)
	at org.apache.spark.sql.catalyst.util.FailureSafeParser.parse(FailureSafeParser.scala:60)
	... 29 more
Caused by: java.lang.RuntimeException: Malformed CSV record
	... 34 more
...

scala> spark.conf.set("spark.sql.csv.parser.columnPruning.enabled", "true")

scala> spark.read.option("header", "true").option("mode", "FAILFAST").csv("fruit.csv").select('fruit, 'color).show()
+------+------+
| fruit| color|
+------+------+
| apple|   red|
|banana|yellow|
|orange|orange|
|   xxx|  null|
+------+------+
```

Comment (viirya) [reply python/pyspark/sql/readwriter.py]: Yeah, I think so. Will move it to under `mode` parameter.

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master and branch-2.4

**Pull request closed without merging**